sql/core/src/test/java/test/org/apache/spark/sql/sources/v2/JavaSimpleWritableDataSource.java

```diff
@@ -0,0 +1,297 @@
+  private StructType schema = new StructType().add("i", "long").add("j", "long");
+
+  class Reader implements DataSourceV2Reader {
+    private String path;
+
+    Reader(String path) {
+      this.path = path;
+    }
+
+    @Override
+    public StructType readSchema() {
+      return schema;
+    }
+
+    @Override
+    public List<ReadTask<Row>> createReadTasks() {
+      return java.util.Arrays.asList(new JavaSimpleCSVReadTask(path));
+    }
+  }
+
+  static class JavaSimpleCSVReadTask implements ReadTask<Row>, DataReader<Row> {
+    private String path;
+    private volatile Iterator<String> lines;
+    private volatile String currentLine;
+
+    JavaSimpleCSVReadTask(Iterator<String> lines) {
+      this.lines = lines;
+    }
+
+    JavaSimpleCSVReadTask(String path) {
+      this.path = path;
+    }
+
+    @Override
+    public DataReader<Row> createReader() {
+      assert path != null;
+      try {
+        if (Files.exists(Paths.get(path))) {
+          return new JavaSimpleCSVReadTask(Files.readAllLines(Paths.get(path)).iterator());
+        } else {
+          return new JavaSimpleCSVReadTask(Collections.emptyIterator());
+        }
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public boolean next() {
+      if (lines.hasNext()) {
+        currentLine = lines.next();
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    @Override
+    public Row get() {
+      String[] values = currentLine.split(",");
+      assert values.length == 2;
+      long l1 = Long.valueOf(values[0]);
+      long l2 = Long.valueOf(values[1]);
+      return new GenericRow(new Object[] {l1, l2});
+    }
+
+    @Override
+    public void close() throws IOException {
+
+    }
+  }
+
+  @Override
+  public DataSourceV2Reader createReader(DataSourceV2Options options) {
+    return new Reader(options.get("path").get());
+  }
+
+
+  class Writer implements DataSourceV2Writer {
+    private String path;
+
+    Writer(String path) {
+      this.path = path;
+    }
+
+    @Override
+    public DataWriteFactory<Row> createWriterFactory() {
+      return new JavaSimpleCSVDataWriteFactory(path);
+    }
+
+    @Override
+    public void commit(WriterCommitMessage[] messages) {
+
+    }
+
+    @Override
+    public void abort() {
+
+    }
+  }
+
+  class InternalRowWriter implements DataSourceV2Writer, SupportsWriteInternalRow {
+    private String path;
+
+    InternalRowWriter(String path) {
+      this.path = path;
+    }
+
+    @Override
+    public DataWriteFactory<InternalRow> createInternalRowWriterFactory() {
+      return new JavaInternalRowCSVDataWriteFactory(path);
+    }
+
+    @Override
+    public void commit(WriterCommitMessage[] messages) {
+
+    }
+
+    @Override
+    public void abort() {
+
+    }
+  }
+
+  static class JavaSimpleCSVDataWriteFactory implements DataWriteFactory<Row>, DataWriter<Row> {
```

Comment:
IF these test classes worked with a Hadoop FS, they'd not just be testable against the local file:// FS, you could bring up a miniDFS cluster and see how that worked. Why? Gives a broader test of how well the API worked against HDFS and the various failure modes, but also of how you pass configuration down.